RFC Draft.md

```diff
@@ -0,0 +1,583 @@
+downloadBlob(downloadResult.content, 'download.jpg');
+```
+
+## Changes to the `put` return object
+
+To better capture customer intent the `put` API will be renamed to `upload`. Additionally `upload` will enable resumability by default in order to simplify API usage and remove the need to provide callbacks for monitoring upload status in favor of a Promise.
+
+**Amplify v5 (`aws-amplify@5`)**
+
+```TypeScript
+// Upload a public file with resumability enabled
+const uploadTask = Storage.put('movie.mp4', fileBlob, {
+  resumable: true,
+  level: 'public',
+  progressCallback: (progress) => {
+    // Progress of upload
+  },
+  completeCallback: (event) => {
+    // Upload finished
+  },
+  errorCallback: (err) => {
+    // Upload failed
+  }
+});
+
+// Pause & resume upload
+uploadTask.pause();
+uploadTask.resume();
+```
+
+**Proposed Amplify v6 (`aws-amplify@6`)**
+
+```TypeScript
+// Upload a public file with resumability enabled by default
+const uploadTask = Storage.put({
+  key: getObjectReference('movie.mpg', { level: 'public' }),
+  content: fileBlob
+});
+
+// Pause & resume upload
+let currentTransferStatus = uploadTask.pause();
+currentTransferStatus = uploadTask.resume();
+
+// Get the current progress of the upload
+const currentTransferProgress = uploadTask.getProgress();
+
+// Wait for the upload to finish (or fail)
+const uploadedObjectReference = await uploadTask.result;
+```
+
+Try out the new `storage` types here: https://www.typescriptlang.org/play#gist/292f4e24178bfca5881aa20961b930dc
+
+# `API` Category Changes
+Amplify is proposing the following changes for the `API` category.
+
+## First param is an object with named parameters
+To improve the readability of our APIs we will be introducing an object parameter to capture request parameters.
+**Amplify v5 (`aws-amplify@5`)**
+
+```typescript
+const apiName = 'MyApiName';
+const path = '/path';
+const myInit = {
+  headers: {}, // OPTIONAL
+  response: true, // OPTIONAL (return the entire Axios response object instead of only response.data)
+  queryStringParameters: {
+    name: 'param' // OPTIONAL
+  }
+};
+
+API.get(apiName, path, myInit)
+  .then((response) => {
+    // Add your code here
+  })
+  .catch((error) => {
+    console.log(error.response);
+  });
+```
+**Proposed Amplify v6 (`aws-amplify@6`)**
+```typescript
+API.get({
+  apiName: 'MyApi',
+  path: '/items',
+  authMode: 'AWS_IAM'
+}, myInit)
+  .then((result) => {
+    // Add your code here
+  }).catch((error) => {
+    console.log(error.response);
+  });
+```
+
+## Adding TypeScript generics to request body and response
+To improve developer experience and permit more strict typing we will be adding generic support to our `API` category APIs.
+
+**Amplify v5 (`aws-amplify@5`)**
+
+Amplify v5 does not support using generics for the request body or response.
+
+**Proposed Amplify v6 (`aws-amplify@6`)**
+
+```typescript
+type MyApiResponse = { firstName: string, lastName: string };
+
+API.get<MyApiResponse>({
```

Comment:
Is this actually better? Are there any guarantees given by passing these as generics that are an improvement over casting the result `const res = result as MyApiResponse` down in use?

If we aren't doing anything to guarantee these types are what is returned, I think there's a case to be made that this is worse than making them unknown and having the customer do type narrowing in their code.